src/compute/src/communication.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Code to spin up communication mesh for a compute replica.
+//! Most of this logic was taken from Timely's communication::networking
+//! module, with the crucial difference that we
+//! actually verify that we can receive data on a connection before considering it
+//! established, and retry otherwise. This allows us to cope with the phenomenon
+//! of spuriously accepted connections that has been observed in Kubernetes with linkerd.
+
+use std::any::Any;
+use std::io::{self, Write};
+use std::io::{Read, Result as IoResult};
+use std::net::{TcpListener, TcpStream};
+use std::sync::Arc;
+use std::thread;
+use std::thread::sleep;
+use std::time::Duration;
+
+use timely::communication::allocator::zero_copy::initialize::initialize_networking_from_sockets;
+use timely::communication::allocator::GenericBuilder;
+use timely::communication::{initialize_from, WorkerGuards};
+use timely::worker::Worker;
+use timely::{CommunicationConfig, Config};
+
+// This constant is sent along immediately after establishing a TCP stream, so
+// that it is easy to sniff out Timely traffic when it is multiplexed with
+// other traffic on the same port.
+const HANDSHAKE_MAGIC: u64 = 0xc2f1fb770118add9;
+
+// This constant is sent upon receiving a connection. The process establishing
+// the connection must receive it in order to consider the connection to be
+// successfully established.
+const ACK_MAGIC: u64 = 0x68656c6f77726c64;
+
+/// Fork of timely::execute::execute that calls into our fork of create_sockets
+pub fn execute_acking_connections<T, F>(
+    mut config: Config,
+    func: F,
+) -> Result<WorkerGuards<T>, String>
+where
```

Comment:
It may be that much of this can be elided and the use can be replaced by `execute_from`, which allows you to show up with your own `Vec<A> where A: AllocateBuilder` and it will take things from there. It would mean we would need to form the `Vec<A>`, but I think we do plan to do that (it is one of the basic blocks down below). Anyhow, it might tidy things up, and reduce the incidental complexity of the PR!